### PR TITLE
Hitvec.2

### PR DESCRIPTION
**Pull request by hollasch**
Sorry about this change; it's pretty much a squash of my changes, after chasing a bunch of dead ends. Should have been more incremental and disciplined. Still, I'm happy with the final results.

The primary change in this is that we're now using `hittable_list` with a `std::vector<hittable*>`, which alleviates a bunch of the awkward memory allocation. This opened up a number of opportunities to simplify code.

This version still leaks memory, but would moving to shared pointers from this base would be straight-forward.

Related/motivating issues: #305, #158, #147

### CODE REVIEW

File: src/TheNextWeek/main.cc [hunk 1]
```diff
@@ -20,272 +20,307 @@
...
     hit_record rec;
-    if (depth <= 0 || !world->hit(r, 0.001, infinity, rec))
+
+    // If we've exceeded the ray bounce limit, no more light is gathered.
+    if (depth <= 0)
         return vec3(0,0,0);
 
+    // If the ray hits nothing, return the background color.
+    if (!world.hit(r, 0.001, infinity, rec))
+        return vec3(0.5,0.5,0.5);
```
Comment (trevordblack) [inline src/TheNextWeek/main.cc]: I understand why this change exists (having grey background).
I just don't love it, and having a scene-specific background was a big reason to create a dedicated `scene` class (for me).

At least as an intermediate step, create a `scene_background` function that is `inline`d and returns vec3(0.5)

File: src/TheNextWeek/hittable_list.h [hunk 2]
```diff
@@ -11,34 +11,56 @@
...
-    if (list_size < 1) return false;
+    if (objects.empty()) return false;
 
     aabb temp_box;
-    bool first_true = list[0]->bounding_box(t0, t1, temp_box);
+    bool first_true = objects[0]->bounding_box(t0, t1, temp_box);
 
     if (!first_true)
         return false;
     else
```
Comment (trevordblack) [inline src/TheNextWeek/hittable_list.h]: `else` is unnecessary here. Remove.

File: src/TheRestOfYourLife/hittable_list.h [hunk 3]
```diff
@@ -11,50 +11,58 @@
...
-    if (list_size < 1) return false;
+    if (objects.empty()) return false;
 
     aabb temp_box;
-    bool first_true = list[0]->bounding_box(t0, t1, temp_box);
+    bool first_true = objects[0]->bounding_box(t0, t1, temp_box);
 
     if (!first_true)
         return false;
     else
```
Comment (trevordblack) [inline src/TheRestOfYourLife/hittable_list.h]: `else` is unneccessary here.
Remove

File: src/TheRestOfYourLife/hittable_list.h [hunk 4]
```diff
@@ -63,19 +71,23 @@ bool hittable_list::bounding_box(double t0, double t1, aabb& output_box) const {
...
+    for (auto object : objects)
+        sum += weight * object->pdf_value(o, v);
+
+    return sum;
 }
 
+
+vec3 hittable_list::random(const vec3 &o) const {
+    auto int_size = static_cast<int>(objects.size());
+    auto index = static_cast<size_t>(random_int(0, int_size-1));
```
Comment (trevordblack) [inline src/TheRestOfYourLife/hittable_list.h]: why the cast back to `size_t`?

File: src/TheRestOfYourLife/main.cc [hunk 5]
```diff
@@ -16,24 +16,31 @@
...
     hit_record rec;
-    if (depth <= 0 || !world->hit(r, 0.001, infinity, rec))
+
+    // If we've exceeded the ray bounce limit, no more light is gathered.
+    if (depth <= 0)
         return vec3(0,0,0);
 
+    // If the ray hits nothing, return the background color.
+    if (!world.hit(r, 0.001, infinity, rec))
+        return vec3(0,1,1);
```
Comment (trevordblack) [inline src/TheRestOfYourLife/main.cc]: Consider creating a `background_color` function

File: src/TheNextWeek/bvh.h [hunk 6]
```diff
@@ -13,12 +13,20 @@
...
 #include "common/rtweekend.h"
 #include "hittable.h"
+#include <algorithm>
 
 
 class bvh_node : public hittable  {
     public:
-        bvh_node() {}
-        bvh_node(hittable **l, int n, double time0, double time1);
+        bvh_node() = delete;
```
Comment (trevordblack) [inline src/TheNextWeek/bvh.h]: That's going to be a Yikes from me.

I think that deleting the default constructor is a good idea.
I just hate the committee's choice of syntax.
I think it's best to gloss over this detail, and lead it to the user to grokk what's happening

File: src/TheNextWeek/bvh.h [hunk 7]
```diff
@@ -13,12 +13,20 @@
...
+
+        bvh_node::bvh_node(hittable_list& list, double time0, double time1)
```
Comment (trevordblack) [inline src/TheNextWeek/bvh.h]: This threw an error for me in clang
`bvh_node::bvh_node` -> `bvh_node`

It's complaining about the scope

File: src/TheNextWeek/bvh.h (hunk 6 above)
Comment (trevordblack) [reply src/TheNextWeek/bvh.h]: Ehhh, I guess the syntax is fine.

I guess the `= delete;` activated my "this is outside our usual c++ calls" lizard brain.

File: src/TheNextWeek/bvh.h [hunk 8]
```diff
@@ -28,95 +36,77 @@ class bvh_node : public hittable  {
...
-}
-
-int box_y_compare (const void * a, const void * b)
-{
-    aabb box_left, box_right;
-    hittable *ah = *(hittable**)a;
-    hittable *bh = *(hittable**)b;
+bool box_x_compare (const hittable* a, const hittable* b) { return box_compare(a, b, 0); }
+bool box_y_compare (const hittable* a, const hittable* b) { return box_compare(a, b, 1); }
+bool box_z_compare (const hittable* a, const hittable* b) { return box_compare(a, b, 2); }
```
Comment (trevordblack) [inline src/TheNextWeek/bvh.h]: hops through 2 function redirections.

Consider inlining `box_compare`

File: src/TheNextWeek/bvh.h [hunk 9]
```diff
@@ -28,95 +36,77 @@ class bvh_node : public hittable  {
...
-    if (box_left.min().y() - box_right.min().y() < 0.0)
-        return -1;
-    else
-        return 1;
-}
+bvh_node::bvh_node(
+    std::vector<hittable*>& objects, size_t start, size_t end, double time0, double time1
+) {
+    int axis = random_int(0,2);
+    auto comparator = (axis == 0) ? box_x_compare
```
Comment (trevordblack) [inline src/TheNextWeek/bvh.h]: Consider replacing with a switch-case or if-elseif

Comment (trevordblack) [reply src/TheNextWeek/bvh.h]: ... instead of the ternary command

File: src/TheNextWeek/bvh.h (hunk 8 above)
Comment (trevordblack) [reply src/TheNextWeek/bvh.h]: ... Actually, don't consider it. _Just do it_

Considering that std::sort is a **significant** part of our runtime, unless the compiler is automatically inlining the function, there's a good 2-5% just on the floor here.

and the
```
if (!a->bounding_box(0,0, box_a) || !b->bounding_box(0,0, box_b))
        std::cerr << "No bounding box in bvh_node constructor.\n";
```
is easily another drag on perf (but I don't recommend removing it)

File: src/TheRestOfYourLife/bvh.h [hunk 10]
```diff
@@ -13,12 +13,20 @@
...
+#include <algorithm>
 
 
 class bvh_node : public hittable  {
     public:
-        bvh_node() {}
-        bvh_node(hittable **l, int n, double time0, double time1);
+        bvh_node() = delete;
+
+        bvh_node::bvh_node(hittable_list &list, double time0, double time1)
```
Comment (trevordblack) [inline src/TheRestOfYourLife/bvh.h]: throws error in clang

File: src/TheRestOfYourLife/bvh.h [hunk 11]
```diff
@@ -28,131 +36,82 @@ class bvh_node : public hittable  {
...
 
-int box_y_compare (const void * a, const void * b)
-{
-    aabb box_left, box_right;
-    hittable *ah = *(hittable**)a;
-    hittable *bh = *(hittable**)b;
+bvh_node::bvh_node(
+    std::vector<hittable*> &objects, size_t start, size_t end, double time0, double time1
+) {
+    if (end <= start) {
```
Comment (trevordblack) [inline src/TheRestOfYourLife/bvh.h]: Should this be copied over to _TheNextWeek_?

File: books/RayTracingInOneWeekend.html [hunk 12]
```diff
@@ -2307,38 +2323,45 @@
...
+
+        objects.add(new sphere(vec3(0,-1000,0), 1000, new lambertian(vec3(0.5, 0.5, 0.5))));
+
         int i = 1;
         for (int a = -11; a < 11; a++) {
             for (int b = -11; b < 11; b++) {
                 auto choose_mat = random_double();
                 vec3 center(a + 0.9*random_double(), 0.2, b + 0.9*random_double());
-                if ((center-vec3(4,0.2,0)).length() > 0.9) {
+                if ((center - vec3(4, .2, 0)).length() > 0.9) {
```
Comment (trevordblack) [inline books/RayTracingInOneWeekend.html]: `vec3(4, 0.2, 0))`

File: books/RayTracingTheNextWeek.html [hunk 13]
```diff
@@ -641,20 +642,23 @@
...
-        if (list_size < 1) return false;
+        if (objects.empty()) return false;
+
         aabb temp_box;
-        bool first_true = list[0]->bounding_box(t0, t1, temp_box);
+        bool first_true = objects[0]->bounding_box(t0, t1, temp_box);
+
         if (!first_true)
             return false;
         else
```
Comment (trevordblack) [inline books/RayTracingTheNextWeek.html]: unnecessary else

File: books/RayTracingTheNextWeek.html [hunk 14]
```diff
@@ -756,35 +766,38 @@
...
+        } else if (object_span == 2) {
+            if (comparator(objects[start], objects[start+1])) {
+                left = objects[start];
+                right = objects[start+1];
+            } else {
+                left = objects[start+1];
+                right = objects[start];
+            }
+        } else {
+            std::sort(objects.begin() + start, objects.begin() + end, comparator);
```
Comment (trevordblack) [inline books/RayTracingTheNextWeek.html]: Need to add `#include <algorithm>` for the `std::sort`

File: books/RayTracingTheNextWeek.html [hunk 15]
```diff
@@ -1825,21 +1858,21 @@
...
+        auto white = new lambertian(new constant_texture(vec3(0.73, 0.73, 0.73)));
+        auto green = new lambertian(new constant_texture(vec3(0.12, 0.45, 0.15)));
+        auto light = new diffuse_light(new constant_texture(vec3(15, 15, 15)));
 
-        return new hittable_list(list,i);
+        objects.add(new yz_rect(0, 555, 0, 555, 555, green));
+        objects.add(new yz_rect(0, 555, 0, 555, 0, red));
+        objects.add(new xz_rect(213, 343, 227, 332, 554, light));
+        objects.add(new xz_rect(0, 555, 0, 555, 0, white));
+        objects.add(new xy_rect(0, 555, 0, 555, 555, white));
```
Comment (trevordblack) [inline books/RayTracingTheNextWeek.html]: The sharing of materials between the objects in this scene is the #1 reason that rewriting the memory deletions is so hard
:((

File: src/TheNextWeek/main.cc (hunk 1 above)
Comment (hollasch) [reply src/TheNextWeek/main.cc]: 1. Make the change easy.
2. Make the easy change.

First step is to have code for a background color. Next step is to use it in the `ray_color` function. Final step is to get the background color from the scene/render-params. This is just the first iteration.

File: src/TheNextWeek/hittable_list.h (hunk 2 above)
Comment (hollasch) [reply src/TheNextWeek/hittable_list.h]: Will fix.

File: src/TheRestOfYourLife/hittable_list.h (hunk 3 above)
Comment (hollasch) [reply src/TheRestOfYourLife/hittable_list.h]: Done.

File: src/TheRestOfYourLife/hittable_list.h (hunk 4 above)
Comment (hollasch) [reply src/TheRestOfYourLife/hittable_list.h]: Just being explicit about the conversion from signed to unsigned. But it looks like my compiler doesn't care either, so I'll remove the cast.

File: src/TheRestOfYourLife/main.cc (hunk 5 above)
Comment (hollasch) [reply src/TheRestOfYourLife/main.cc]: Future attractions.

File: src/TheNextWeek/bvh.h (hunk 6 above)
Comment (hollasch) [reply src/TheNextWeek/bvh.h]: I was being pedantic about disallowing the construction of unspecified BVH nodes, but I can definitely see how I'm just adding a cognitive speed bump in this context. Would totally do this in production code, but will simplify.

Comment (hollasch) [reply src/TheNextWeek/bvh.h]: I'm just going to pull the parameterless ctor altogether, and ignore the fact that C++ will autogenerate the default empty constructor.

File: src/TheNextWeek/bvh.h (hunk 7 above)
Comment (hollasch) [reply src/TheNextWeek/bvh.h]: All gone.

File: src/TheNextWeek/bvh.h (hunk 8 above)
Comment (hollasch) [reply src/TheNextWeek/bvh.h]: Eh. Ok.

File: src/TheNextWeek/bvh.h (hunk 9 above)
Comment (hollasch) [reply src/TheNextWeek/bvh.h]: So, you want me to use modern C++ here? `decltype` or `std::function<>`?  :D

Comment (hollasch) [reply src/TheNextWeek/bvh.h]: Hmmm. The two alternatives are

```
    decltype(box_x_compare) *comparator;
    switch (random_int(0,2)) {
        default:
        case 0: comparator = box_x_compare; break;
        case 1: comparator = box_y_compare; break;
        case 2: comparator = box_z_compare; break;
    }
```
or
```
    #include <functional>
    ...
    std::function<bool(const hittable*, const hittable*)> comparator;
    switch (random_int(0,2)) {
        default:
        case 0: comparator = box_x_compare; break;
        case 1: comparator = box_y_compare; break;
        case 2: comparator = box_z_compare; break;
    }
```
Gonna stick with ternary; I believe it's much easier for a general audience to grok.

File: src/TheRestOfYourLife/bvh.h (hunk 10 above)
Comment (hollasch) [reply src/TheRestOfYourLife/bvh.h]: All gone.

File: src/TheRestOfYourLife/bvh.h (hunk 11 above)
Comment (hollasch) [reply src/TheRestOfYourLife/bvh.h]: _Oif_. Excellent catch.

File: books/RayTracingInOneWeekend.html (hunk 12 above)
Comment (hollasch) [reply books/RayTracingInOneWeekend.html]: Will do.

File: books/RayTracingTheNextWeek.html (hunk 13 above)
Comment (hollasch) [reply books/RayTracingTheNextWeek.html]: Done.

File: src/TheNextWeek/bvh.h (hunk 8 above)
Comment (hollasch) [reply src/TheNextWeek/bvh.h]: Actually, I can hoist this sanity check out as a prepass on the unsorted `std::vector` before invoking the sort function.

File: books/RayTracingTheNextWeek.html (hunk 15 above)
Comment (hollasch) [reply books/RayTracingTheNextWeek.html]: This change is largely about moving the dynamic list to using `std::vector`, and thus having defined destructor logic. It currently has bare `hittable` pointers. For the next rev, we can convert this to `std::vector<std::shared_ptr<hittable>>`, and then ensure that object/material creation returns a shared pointer.

File: src/TheNextWeek/bvh.h (hunk 8 above)
Comment (hollasch) [reply src/TheNextWeek/bvh.h]: Oh, never mind; it's not a sanity check, it actually computes the aabb (over and over; sigh).

I actually did start out on a cached aabb implementation, but that ended up in the weeds pretty quickly. Rolled back for a more incremental approach.

**Comment by hollasch** [comment]
Block of 7 commits above incorporate review feedback.
